simexpal/queuesock.py

```diff
@@ -1,105 +1,121 @@
-
-class _Connection:
-	@staticmethod
-	def handle_sock(self, descriptor):
-		assert self.state == _State.REQUEST
-
-		data = self.sock.recv(4096)
-		if data:
-			self.recv_buffer += data
-			return
+class Queue:
 
-		req = util.yaml_from_string(self.recv_buffer.decode())
-		resp = self.queue.dispatch(descriptor.get_loop(), req)
-		self.state = _State.DONE
-		self._handle.unregister()
+	def __init__(self, serve_socket):
+		self.socket = serve_socket
+		self.socket_path = serve_socket.getsockname()
+		self.selector = selectors.DefaultSelector()
+		self.subprocess = None
+		self.subprocess_terminated = True
+		self.requests = []
 
 	@staticmethod
-	def observe_loop(self, descriptor):
-		self._handle.unregister()
-		self._observer.unregister()
-
-	def __init__(self, queue, sock):
-		self.queue = queue
-		self.sock = sock
-		self.recv_buffer = bytes()
-		self.state = _State.REQUEST
-		self._handle = None
-		self._observer = None
-
-	def run(self, loop):
-		self._handle = loop.register_file(self.sock, evloop.READ,
-				functools.partial(_Connection.handle_sock, self))
-		self._observer = loop.register_observer(functools.partial(_Connection.observe_loop, self))
-
-def run_queue(loop, sockfd=None, force=False):
+	def get_display_name(spec_yml):
+		display_name = spec_yml['manifest']['experiment']
+
+		variants = spec_yml['manifest']['variants']
+		if variants:
+			display_name += ' ~ ' + ', '.join([variant['name'] for variant in variants])
+
+		revision = spec_yml['manifest']['revision']
+		if revision:
+			display_name += ' @ ' + revision
+
+		return display_name
+
+	def run(self):
+		print('Serving on {}'.format(self.socket_path))
+		self.socket.listen()
+
+		while True:
+			connection = Connection(self.socket)
+			try:
+				self.selector.register(connection.socket, selectors.EVENT_READ, connection)
+			# The socket is already registered, due to a non-action loop.
+			except KeyError:
+				pass
+
+			events = self.selector.select(1)
+
+			for sk, mask in events:
+				data = sk.data.progress()
+				if data:
+					self.requests.append(data)
+				sk.data.close()
+				self.selector.unregister(sk.fd)
+
+			try:
+				self.subprocess_terminated = self.subprocess.poll() is not None
+			# No subprocess has been started yet.
+			except AttributeError:
+				pass
+
+			if self.subprocess_terminated and not len(self.requests) == 0:
+				request = self.requests.pop(0)
+
+				if request['action'] == 'launch':
+					display_name = self.get_display_name(request)
+					print("Launching run {}/{}[{}]".format(
+						display_name, request['manifest']['instance'], request['manifest']['repetition']))
+
+					script = os.path.abspath(sys.argv[0])
+					specfile = request['specfile_path']
+
+					self.subprocess = subprocess.Popen([script, 'internal-invoke', '--queue', specfile])
+
+				elif request['action'] == 'stop':
+					print("Closing socket on {}".format(self.socket_path))
+					self.socket.shutdown(socket.SHUT_RDWR)
+					self.socket.close()
```

Comment:
shutdown() before close() is not necessary.

Reply:
Done